docs/spec/channel.md

```diff
@@ -297,9 +297,12 @@ queueing requests (e.g. health checks) are not constrained.
 The HTTP event queueing request's URL MUST correspond to a single, unique
 Channel at any given moment in time. This MAY be done via the host, path, query
 string, or any combination of these. This mapping is handled exclusively by the
-Channel implementation, exposed via the Channel's `status.address`. If an HTTP
-event queueing request's URL does not correspond to an existing Channel, then
-the Channel MUST respond with `404 Not Found`.
+Channel implementation, exposed via the Channel's `status.address`.
+
+If an HTTP event queueing request's URL does not correspond to an existing
+Channel, then the failure behaviour is undefined. For example, the Channel MAY
```

Comment:
I wouldn't say it's undefined. We want the client to receive an error response in some form.

> If an HTTP event queueing request's URL does not correspond to an existing
Channel, the requesting client MUST receive an error response. The error response may differ depending on implementation. Examples include HTTP `404 Not Found` status, DNS resolution failure, or connection refused.

Reply:
well it's really undefined because whatever the URL corresponds to might accept the event. Maybe I'm confused by "an existing Channel".

Reply:
@grantr Let's assume you try to send a request to a non existing channel like `foo-kn-channel.default.svc` in case of IMC (or Kafka) channel implementations. There are two situations:

1. You get a DNS error like "unable to resolve hostname foo"
2. If the channel doesn't exist anymore *but* existed in past and the sender already made a request to it, it has the DNS record cached so it will return a 404 (because the imc-dispatcher receives the request, doesn't recognize the `Host` header and replies with 404)
3. If the imc-dispatcher service is not up *but* existed in past and the sender already made a request to it, it has the DNS record cached so it will return an i/o timeout, because the imc-dispatcher is not available

To me these looks situations specific to these two implementations. In case of a channel where 1 dispatcher deployment = 1 channel CR, the case 2 might be different: you could get an i/o timeout, because the deployment doesn't exist anymore.

What I'm saying is: let's avoid to define a specific behaviour, because it may be problematic to implement depending on the chosen architecture. That's proven by the fact that even our reference implementation doesn't fulfil this 404 contract completely.
If a channel is not available/not existing, it's up to the implementation to decide how to signal it to the user, as soon as the signal is some kind of error (404, i/o timeout, DNS error, etc) and the message doesn't go through

Reply:
Can't we just keep it a bit more lenient, so that we have the channel implementation to decide what kind of "error" to report ?

Reply:
I'm fine with relaxing the requirements, I'm just feeling like we're mixing and matching what's under the channels control and what is not. For example:

```
If a channel is not available/not existing, it's up to the implementation to decide how to signal it to the user, as soon as the signal is some kind of error (404, i/o timeout, DNS error, etc) and the message doesn't go through
```

If I use curl to a an address given in the status.address, channel implementor is not in any part of that (well, except lying about the location of the channel in its status.address).

Reply:
How do you propose to change it @vaikas ?

Reply:
First I wanted to make sure my understanding was correct :) If for some reason the client making the call fails to connect to the specified endpoint (dns, network partition, whatever) seems like most certainly those categories of errors are not defined and in some cases might not be surfaced to the caller in the correct manner (poor libraries, etc.).
So, I think we can constrain this spec starting with the assumption that the channel is actually receiving the HTTP request, otherwise we'd be in similar predicament as with the method not allowed above (405), correct? So, if that's indeed what we're tackling here, it does seem reasonable to me that we expect the channel to return 404 in that case if the endpoint has not been correctly configured?
So I think we're back to status.address being wrong (which is a bug), or the fact that the status is out of date with the provisioning (which is a bug). Both of these cases might lead to network related issues you were describing, but they are bugs and I don't think we should relaxing the spec to work around bad implementations.
If there's a channel implementation that handles multichannels by path portion, say:
channel.example.com/namespace/name
and the status.address for default/mychannel would then be:
channel.example.com/default/mychannel
If I now were to make a request to:
channel.example.com/default/mychannelnotthere
It would be perfectly reasonable to expect that to return a 404 and that should be totally doable from the channels implementors point of view?
Note that I could still make a call to either of those URLs and they could indeed fail at any given time at the network level where the Channel does not even receive the request even for a Channel that exists.

Reply:
> So, I think we can constrain this spec starting with the assumption that the channel is actually receiving the HTTP request,

What I'm saying is that we cannot do to that assumption. If you send a request to a channel that does not exist, it means you're sending a request to `channel-x-kn-channel.mynamespace.svc.cluster.local`, which won't be resolved by the DNS because no `ExternalService` is registered for that.

> It would be perfectly reasonable to expect that to return a 404 and that should be totally doable from the channels implementors point of view?

But that's how broker works today, not channels. The channels works using `ExternalService` k8s svcs, so when you perform the request you perform it to a precise dns name, which is always resolved in the same service name (eg `channel-x-kn-channel.mynamespace.svc.cluster.local`, `channel-y-kn-channel.mynamespace.svc.cluster.local`, ... they always resolve to `imc-dispatcher.knative-eventing.svc.cluster.local`). The channels always receives the requests at the path `/` and, looking at the `Host` header, they recognize at which the request was made. If that DNS entry doesn't exist, you won't reach at all the imc-dispatcher, hence you'll get some DNS failure, nor a 404

Reply:
If I'm following, I agree that from a client's perspective the error they see might not be a 404, it could technically be just about anything. However, we have no control over things that do not reach our code. Or put another way, the spec can only dictate what happens once a request reaches our code and at that point it seems fair to mandate it's a 404.

Perhaps what @slinkydeveloper is looking for is something more akin to what might appear in a Primer document - something that provides guidance to clients on how they might react to different types of errors. Not sure we can mandate what they do with RFC2119 language, but hints could be useful.

Reply:
Ping @duglin @vaikas @lionelvillard @grantr @matzew  this is still an issue, any ideas how we want to proceed forward?

Reply:
Rereading this... what if we rephrased it so that it's more focused on what a channel should do, rather than what a client will see. For example, something like:
```
If a channel implementation receives a request that does not correspond to a known channel, then a
'404 Not Found' MUST be returned.
```

Reply:
@duglin sounds reasonable to me, this way the behaviour of non existing channel from the client point of view remains  "undefined" under certain circumstances, aka the channel guarantees that replies with a 404 only if it actually received the request